### PR TITLE
CBG-724: Deletions of empty documents are not imported

### PR DESCRIPTION
**Pull request by sarathkumarsivan**

### CODE REVIEW

File: db/document.go [hunk 1]
```diff
@@ -525,7 +528,8 @@ func (doc *Document) IsSGWrite(rawBody []byte) (isSGWrite bool, crc32Match bool)
 		base.Warnf("Unable to marshal doc body during SG write check for doc %s. Error: %v", base.UD(doc.ID), err)
 		return false, false
 	}
-	if base.Crc32cHashString(bodyBytes) == doc.SyncData.Crc32c {
+
+	if base.Crc32cHashString(bodyBytes) == doc.SyncData.Crc32c && emptyBodyCrc32cHash != doc.SyncData.Crc32c {
```
Comment (adamcfraser) [inline db/document.go]: This approach isn't valid - it doesn't handle the following case:
1. Empty document created
2. Non-body mutation made to empty document (eventing xattr, expiry, etc)

In this scenario, isSGWrite should be returning (true, true).  With the above clause, it's going return (false,false) and trigger an unwanted import.

**Review by adamcfraser — CHANGES_REQUESTED** [review]
The proposed approach doesn't address the underlying issue.  Were you able to identify why a delete mutation is coming through with a body of '{}' in the first place?

File: db/document.go (hunk 1 above)
Comment (sarathkumarsivan) [reply db/document.go]: Changes are committed to handle this scenario in the latest revision.

**Review by sarathkumarsivan — COMMENTED** [review]
When the document is a delete, the bodyBytes would be replaced with “{}” and we can’t use bodyBytes in CRC-32 checksum comparison to determine whether the mutation was an SG write and the document has already been imported. Need to set the current value of CRC-32 checksum from bodyBytes when the document is non-delete and revert back to the correct CRC-32 checksum “0x00” before we replace the bodyBytes otherwise (if the document is delete).

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Changes look good, a few minor cleanup comments.

File: base/bucket_gocb.go [hunk 2]
```diff
@@ -1224,7 +1225,12 @@ func (bucket *CouchbaseBucketGoCB) UpdateXattr(k string, xattrKey string, exp ui
 			UpsertEx(xattrKey, xv, gocb.SubdocFlagXattr).                                                // Update the xattr
 			UpsertEx(xattrCasProperty, "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros) // Stamp the cas on the xattr
 		if bucket.IsSupported(sgbucket.BucketFeatureCrc32cMacroExpansion) {
-			builder.UpsertEx(xattrBodyHashProperty, "${Mutation.value_crc32c}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros) // Stamp the body hash on the xattr
+			// Stamp the body hash on the xattr
+			if isDelete {
+				builder.UpsertEx(xattrBodyHashProperty, XattrMacroCrc32cDelete, gocb.SubdocFlagXattr)
```
[the remark this answers is not in the document]
Comment (sarathkumarsivan) [reply base/bucket_gocb.go]: Added additional unit tests (TestUpdateXattrWithDeleteBodyAndIsDelete, TestUpdateXattrWithDeleteBodyAndIsNotDelete ) in bucket_gocb_test.go to validate the new behaviour of UpdateXattr.

File: base/bucket_gocb.go [hunk 3]
```diff
@@ -46,8 +46,9 @@ const (
 	// so this is set to 1
 	numNodesPersistTo = uint(1)
 
-	xattrMacroCas         = "cas"
-	xattrMacroValueCrc32c = "value_crc32c"
+	xattrMacroCas          = "cas"
+	xattrMacroValueCrc32c  = "value_crc32c"
+	XattrMacroCrc32cDelete = "0x00"
```
[the remark this answers is not in the document]
Comment (sarathkumarsivan) [reply base/bucket_gocb.go]: Make sense; Refactored XattrMacroCrc32cDelete to DeleteCrc32c and added short comment describes what it represents.

File: db/document.go [hunk 4]
```diff
@@ -525,7 +524,15 @@ func (doc *Document) IsSGWrite(rawBody []byte) (isSGWrite bool, crc32Match bool)
 		base.Warnf("Unable to marshal doc body during SG write check for doc %s. Error: %v", base.UD(doc.ID), err)
 		return false, false
 	}
-	if base.Crc32cHashString(bodyBytes) == doc.SyncData.Crc32c {
+	// Underlying problem is that bodyBytes has been replaced with "{}" if the document is a delete
```
[the remark this answers is not in the document]
Comment (sarathkumarsivan) [reply db/document.go]: Removed obsolete comments left over from working session; added a brief description about the situation and need to modify currentBodyCrc32c in the delete case.

**Review by sarathkumarsivan — COMMENTED** [review]
- Added additional unit tests (TestUpdateXattrWithDeleteBodyAndIsDelete, TestUpdateXattrWithDeleteBodyAndIsNotDelete
) in bucket_gocb_test.go to validate the new behaviour of UpdateXattr.
- Refactored XattrMacroCrc32cDelete to DeleteCrc32c and added short comment describes what it represents.
- Removed obsolete comments left over from working session; added a brief description about the situation and need to modify currentBodyCrc32c in the delete case.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**